aggregator.go

```diff
@@ -0,0 +1,139 @@
+package main
+
+import (
+	"encoding/base64"
+	"flag"
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"log"
+	"os"
+	"sync"
+	"time"
+)
+
+func getLatestOutput(instanceId string, sess *session.Session, withLatest bool) string {
+	var consoleInput *ec2.GetConsoleOutputInput
+	if withLatest {
+		consoleInput = &ec2.GetConsoleOutputInput{
+			InstanceId: aws.String(instanceId),
+			Latest:     aws.Bool(true),
+		}
+	} else {
+		consoleInput = &ec2.GetConsoleOutputInput{
+			InstanceId: aws.String(instanceId),
+		}
+	}
+	ec2Svc := ec2.New(sess)
+
+	result, err := ec2Svc.GetConsoleOutput(consoleInput)
+	if err != nil {
+		if err, ok := err.(awserr.Error); ok {
+			// Non Nitro AWS instance types don't support retrieving the latest console logs
+			if err.Code() == "UnsupportedOperation" && withLatest {
+				return getLatestOutput(instanceId, sess, false)
+			} else {
+				log.Println(err.Error())
+			}
+		} else {
+			log.Println(err.Error())
+		}
+		return ""
+	}
+
+	output, err := base64.StdEncoding.DecodeString(*result.Output)
+	if err != nil {
+		log.Printf("Failed decoding console output from base64\n%s\n", err)
+		return ""
+	}
+
+	return string(output)
+}
+
+func findOverlappingEndingIndex(prev, new string) int {
+	var i int
+	prevLen := len(prev)
+	for i = len(new); i > 0; i-- {
+		if i > prevLen {
+			continue
+		}
+		if new[:i] == prev[prevLen-i:] {
+			break
+		}
+	}
+	return i
+}
+
+func aggregateConsoleOutput(wg *sync.WaitGroup, folder, instanceId string, sess *session.Session) {
+	defer wg.Done()
```

Comment:
It maybe does not matter a lot here. As you have nothing to handle
But, in general, it's best to pass context with cancellation callback to this function

And in main - catch signals, invoke cancel on context and test for it inside this function, to exit goroutines in proper manner